Composable image (#3879)
platformsh/platformsh-docs: sites/friday/src/create-apps/app-reference/composable-image.md

With it, packages are backed as values and all of them are built in total isolation.
Which make it possible to install different versions of the same package very easily.
With [Nix](https://nix.dev/reference/glossary#term-Nix), there is no undeclared dependencies in your source code
and if it works locally, it will work on yours.

{{% note title="TODO" %}}
TODO: What is NixOs ? complete the definition as i'm not sure of it :P

Intersting quote from Jerome's internal presentation: https://drive.google.com/file/d/1_w4NwEae3RWt6lYYW_2Er7Fy7zRJ7tac/view
- `One image to rule them all` --> i think this quote needs to be displayed in the page somewhere at the top, isn't it?
{{% /note %}}

If you want to learn more about Nix, please have a look on [Jérôme Vieilledent's presentation](https://www.youtube.com/watch?v=emOt32DVl28) during the Symfony Con Brussels 2023 (our Product Manager of that Composable Image feature).

## Primary application properties

All application configuration takes place in a `{{< vendor/configfile "app" >}}` file, with each application configured
under a unique key beneath the top-level `applications` key.

> For example, it is possible to deploy multiple runtimes in an application container - one PHP, one JavaScript and one
> Python - for the frontend of a deployed site.

Reviewer: ```suggestion
It is possible to use multiple runtimes in a single application container, e.g. PHP, NodeJS and Python, while remaining in control of their versions.
```

> In this case, the unified `{{< vendor/configfile "app" >}}` file would look like:

Reviewer: ```suggestion
For example, the unified `{{< vendor/configfile "app" >}}` file may look like the following:
```

          - pdo_sqlite
      - "nodejs@{{% latest "nodejs" %}}"
      - "python@3.12"
    # Additional frontend configuration
```

The following table presents all properties available at the level just below the unique application name (`frontend`
above).

The column _Set in instance?_ defines whether the given property can be overridden within a `web` or `workers` instance.
To override any part of a property, you have to provide the entire property.

| Name               | Type                                                | Required | Set in instance? | Description                                                                                                                                                                                                                                                      |
|--------------------|-----------------------------------------------------|----------|------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `stack`            | An array of [NixOs runtimes](#stack)                | Yes      | No               | The list of runtime images to use with a specific app. Format: <br>`<runtime>@<version>`.                                                                                                                                                                        |

Reviewer: ```suggestion
| `stack`            | An array of [Nix packages](#stack)                | Yes      | No               | A list of packages from our collection of supported runtimes and/or from NixPkgs.                                                                                                                                                                        |
```

{{% note %}}
Please note that, even if available in [Application reference](/create-apps/app-reference/builtin-image.md) when using the built-in image
of defining images to use in your application container, the  ``type``, ``build``, ``dependencies``, and ``runtime`` keywords
are not supported if you choose to use Composable Image (`stack`).
{{% /note %}}

## Root directory

{{% note title="TODO REMOVE ?" %}}
Does this section can't be removed from here as the [source](#source) exists at the end ?
{{% /note %}}

Some of the properties you can define are relative to your app's root directory.
The root defaults to the root of the repository.

```yaml {configFile="app"}
applications:
  frontend:
    stack: [ "nodejs:{{% latest nodejs %}}" ]
    # Default behavior of source.root
    source:
      root: "/"
    # Additional frontend configuration
```

That is, if a custom value for `source.root` is not provided in your configuration, the default behavior is equivalent
to the above.

To specify another directory, for example for a [multi-app project](../multi-app/_index.md),
use the [`source.root` property](#source).

## Stack

> The ``stack`` keyword is used for defining runtimes and other binaries (like `wkhtmltopdf`, or `imagemagick`, see list
> below),
> as a YAML array, that you would like to install in your application container.

Reviewer: ```suggestion
The ``stack`` keyword is used to define the list of all the packages your application needs to work, 
including language runtimes (e.g. `php`, `nodejs`, `python`...) and other utilities (e.g. `imagemagick`, `graphviz`, `yarn`...).
This list is defined as a YAML array.
```

The version is the major (`X`) and sometimes minor (`X.Y`) version numbers,
depending on the service, as in the following table.

> Security and other patches are taken care of for you automatically.

Reviewer: This is only true for the runtimes.
As for the other packages, it's up to the Nix maintainers.

| [Perl](https://www.perl.org/)                | `perl`        | 5                          |
| [PHP](/languages/php.html)                   | `php`         | 8.3, 8.2, 8.1              |
| [Python](/languages/python.html)             | `python`      | 3.12, 3.11, 3.10, 3.9, 2.7 |

Available tools and their supported versions:

| **Tool**                                                             | **`runtime`** | **Supported `version`** |
|----------------------------------------------------------------------|---------------|-------------------------|
| [Face detector](https://www.thregr.org/~wavexx/software/facedetect/) | `facedetect`  | none                    |
| [GNU Compiler Collection](https://gcc.gnu.org/)                      | `gcc`         | none                    |
| [Graph visualization tools](https://graphviz.org/)                   | `graphviz`    | none                    |
| [ImageMagick](http://www.imagemagick.org/)                           | `imagemagick` | none                    |
| [WkHtmlToPdf](https://wkhtmltopdf.org/)                              | `wkhtmltopdf` | none                    |

> | [Yarn](https://classic.yarnpkg.com/)                                 | `yarn`        | none                    |

Reviewer: I'm not sure we need to specify this.

        disabled_extensions:
          - gd
```

{{% note %}}
To find out the name of the PHP package you want to use, use the ``PHP upstream extension`` value from the [NixOs search engine](https://search.nixos.org/packages?channel=unstable&show=php82Extensions.gd).
If the maintainer of the package does not provide such information (sorry about that), remember that PHP package names on NixOs always respect format ``<PHP><VERSION>Extensions.<EXTENSION-NAME>``,
and you need to copy the ``<EXTENSION-NAME>``.

Example for using [`ZIP` PHP package](https://search.nixos.org/packages?channel=unstable&show=php83Extensions.zip&from=0&type=packages&query=php+zip)
and [`Imagemagick` PHP package](https://search.nixos.org/packages?channel=unstable&show=php83Extensions.imagick&from=0&type=packages&query=php+imagick):
* [`ZIP` PHP package]: value of the ``PHP upstream extension`` is `zip`.
* [`Imagemagick` PHP package]: no value of the ``PHP upstream extension``, use the ``<EXTENSION-NAME>=imagick``.

```yaml {configFile="app"}
applications:
  myapp:
    source:
      root: "/"
    stack:
      - "php@{{% latest "php" %}}":
        extensions:
          - zip
          - imagick
```

> {{% /note %}}

Reviewer: This should probably go to the PHP page

<--->
+++
title=Python
+++
Python packages are installed as new package in your stack, by using the full name of the package.

Example for installing [``python312Packages.yq``](https://search.nixos.org/packages?channel=unstable&show=python312Packages.yq):

```yaml {configFile="app"}
applications:
  myapp:
    stack:
      - "python@3.12"
      - "python312Packages.yq" # python package specific
```

> {{< /codetabs >}}

Reviewer: This should probably go to the Python page

> Alternatively, if you need to include configuration options for your extensions, use either your ``php.ini`` file or [environment variables](/development/variables/set-variables.md):

Reviewer: Should also go to the PHP page
